### PR TITLE
The error encountered while starting a native image

### PR DESCRIPTION
**Pull request by omercelikceng**
Hello,

In my Spring Boot project, I am using Hazelcast version 5.0.0. It was working without any issues, but I decided to create a native image with GraalVM(Spring Boot Project). I successfully created a native image, but I'm encountering an error during the project startup when running the native image.. Because trying to access a field using reflection, normally this field is registered in a hint for us. However,  used reflection to get the field in the code, but you haven't used it. Therefore, this unused field is not registered in the hint. But at runtime, when this field is tried to be accessed using reflection, it causes an error.

This is why I created a pull request to remove this code. Alternatively, if you want me to make improvements in another way, I can update my pull request.

Screenshot 2024-07-30 151829

If you want to reproduce the error, I am sharing a sample POC project.
 https://github.com/omercelikceng/native-poc-cache 

Error : 

```
Caused by: java.lang.ExceptionInInitializerError: null
        at com.hazelcast.spi.properties.ClusterProperty.<clinit>(ClusterProperty.java:1692) ~[na:na]
        at com.hazelcast.client.impl.clientside.HazelcastClientInstanceImpl.<init>(HazelcastClientInstanceImpl.java:232) ~[com.example.NativePocCacheApplication:5.5.0]
        at com.hazelcast.client.HazelcastClient.constructHazelcastClient(HazelcastClient.java:475) ~[com.example.NativePocCacheApplication:5.5.0]
        at com.hazelcast.client.HazelcastClient.newHazelcastClientInternal(HazelcastClient.java:425) ~[com.example.NativePocCacheApplication:5.5.0]
        at com.hazelcast.client.HazelcastClient.newHazelcastClient(HazelcastClient.java:142) ~[com.example.NativePocCacheApplication:5.5.0]
        at com.example.HazelcastConfiguration.hazelcastClient(HazelcastConfiguration.java:43) ~[com.example.NativePocCacheApplication:na]
        at com.example.HazelcastConfiguration$$SpringCGLIB$$0.CGLIB$hazelcastClient$1(<generated>) ~[com.example.NativePocCacheApplication:na]
        at com.example.HazelcastConfiguration$$SpringCGLIB$$FastClass$$1.invoke(<generated>) ~[com.example.NativePocCacheApplication:na]
        at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:258) ~[com.example.NativePocCacheApplication:6.1.4]
        at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331) ~[na:na]
        at com.example.HazelcastConfiguration$$SpringCGLIB$$0.hazelcastClient(<generated>) ~[com.example.NativePocCacheApplication:na]
        at com.example.HazelcastConfiguration__BeanDefinitions.lambda$getHazelcastClientInstanceSupplier$1(HazelcastConfiguration__BeanDefinitions.java:52) ~[na:na]
        at org.springframework.util.function.ThrowingBiFunction.apply(ThrowingBiFunction.java:68) ~[com.example.NativePocCacheApplication:6.1.4]
        at org.springframework.util.function.ThrowingBiFunction.apply(ThrowingBiFunction.java:54) ~[com.example.NativePocCacheApplication:6.1.4]
        at org.springframework.beans.factory.aot.BeanInstanceSupplier.lambda$get$2(BeanInstanceSupplier.java:206) ~[na:na]
        at org.springframework.util.function.ThrowingSupplier.get(ThrowingSupplier.java:58) ~[com.example.NativePocCacheApplication:6.1.4]
        at org.springframework.util.function.ThrowingSupplier.get(ThrowingSupplier.java:46) ~[com.example.NativePocCacheApplication:6.1.4]
        at org.springframework.beans.factory.aot.BeanInstanceSupplier.invokeBeanSupplier(BeanInstanceSupplier.java:218) ~[na:na]
        at org.springframework.beans.factory.aot.BeanInstanceSupplier.get(BeanInstanceSupplier.java:206) ~[na:na]
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.obtainInstanceFromSupplier(DefaultListableBeanFactory.java:949) ~[com.example.NativePocCacheApplication:6.1.4]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.obtainFromSupplier(AbstractAutowireCapableBeanFactory.java:1217) ~[com.example.NativePocCacheApplication:6.1.4]
        ... 33 common frames omitted
        
Caused by: com.hazelcast.core.HazelcastException: java.lang.NoSuchFieldException: ARTIFACT_ID
        at com.hazelcast.instance.BuildInfoProvider.readStaticStringField(BuildInfoProvider.java:112) ~[na:na]
        at com.hazelcast.instance.BuildInfoProvider.readBuildPropertiesClass(BuildInfoProvider.java:92) ~[na:na]
        at com.hazelcast.instance.BuildInfoProvider.getBuildInfoInternalVersion(BuildInfoProvider.java:71) ~[na:na]
        at com.hazelcast.instance.BuildInfoProvider.populateBuildInfoCache(BuildInfoProvider.java:48) ~[na:na]
        at com.hazelcast.instance.BuildInfoProvider.<clinit>(BuildInfoProvider.java:41) ~[na:na]
        ... 54 common frames omitted

Caused by: java.lang.NoSuchFieldException: ARTIFACT_ID
        at java.base@21.0.2/java.lang.Class.checkField(DynamicHub.java:1041) ~[com.example.NativePocCacheApplication:na]
        at java.base@21.0.2/java.lang.Class.getField(DynamicHub.java:1026) ~[com.example.NativePocCacheApplication:na]
        at com.hazelcast.instance.BuildInfoProvider.readStaticStringField(BuildInfoProvider.java:109) ~[na:na]
        ... 58 common frames omitted
```

### CODE REVIEW

**Comment by JackPGreen** [comment]
`artifactId` variable introduced in 5c8bc2c644597f664ee64e3b8b87e1935f2a0439 by me.
I was trying to remove `com.hazelcast.instance.BuildInfoProvider.readBuildPropertiesClass(Class<?>, BuildInfo, Overrides)#distribution` and this was leftover from that - but good spot, it's not needed.

> If you want to reproduce the error, I am sharing a sample POC project.
https://github.com/omercelikceng/native-poc-cache

Thanks for the reproducer, but it seems to work for me locally?

```
[INFO] Successfully built image 'docker.io/library/native-poc-cache:0.0.1-SNAPSHOT'
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:18 min
[INFO] Finished at: 2024-08-01T14:03:10+01:00
[INFO] ------------------------------------------------------------------------
```

Could this be platform-specific?

File: hazelcast/src/main/java/com/hazelcast/instance/BuildInfoProvider.java [hunk 1]
```diff
@@ -26,6 +26,7 @@
 
 /**
  * Provides information about current Hazelcast build.
+ * @author Omer Celik
```
Comment (JackPGreen) [inline hazelcast/src/main/java/com/hazelcast/instance/BuildInfoProvider.java:29]: Nit - many others have worked on this class.

Comment (omercelikceng) [reply hazelcast/src/main/java/com/hazelcast/instance/BuildInfoProvider.java:29]: I thought it was mandatory to write when we make changes, so I'm removing it immediately. I am sorry

**Comment by omercelikceng** [comment]
> `artifactId`[5c8bc2c'de](https://github.com/hazelcast/hazelcast/commit/5c8bc2c644597f664ee64e3b8b87e1935f2a0439) benim tarafımdan tanıtılan değişken . Kaldırmaya çalışıyordum `com.hazelcast.instance.BuildInfoProvider.readBuildPropertiesClass(Class<?>, BuildInfo, Overrides)#distribution`ve bu ondan kalan bir şeydi - ama iyi bir nokta, gerek yok.
> 
> > Hatayı yeniden üretmek istiyorsanız, örnek bir POC projesi paylaşıyorum. 
> > https://github.com/omercelikceng/native-poc-cache
> 
> Üretici için teşekkürler, ancak yerel olarak benim için işe yarıyor gibi görünüyor?
> 
> ```
> [INFO] Successfully built image 'docker.io/library/native-poc-cache:0.0.1-SNAPSHOT'
> [INFO] 
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD SUCCESS
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  04:18 min
> [INFO] Finished at: 2024-08-01T14:03:10+01:00
> [INFO] ------------------------------------------------------------------------
> ```
> 
> Bu durum platforma özgü olabilir mi?

When you run the code it throws an error.

> `artifactId` variable introduced in [5c8bc2c](https://github.com/hazelcast/hazelcast/commit/5c8bc2c644597f664ee64e3b8b87e1935f2a0439) by me. I was trying to remove `com.hazelcast.instance.BuildInfoProvider.readBuildPropertiesClass(Class<?>, BuildInfo, Overrides)#distribution` and this was leftover from that - but good spot, it's not needed.
> 
> > If you want to reproduce the error, I am sharing a sample POC project.
> > https://github.com/omercelikceng/native-poc-cache
> 
> Thanks for the reproducer, but it seems to work for me locally?
> 
> ```
> [INFO] Successfully built image 'docker.io/library/native-poc-cache:0.0.1-SNAPSHOT'
> [INFO] 
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD SUCCESS
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  04:18 min
> [INFO] Finished at: 2024-08-01T14:03:10+01:00
> [INFO] ------------------------------------------------------------------------
> ```
> 
> Could this be platform-specific?

When you run the code, it throws an error. In other words, it throws an error when the project is starting up.

> `artifactId` variable introduced in [5c8bc2c](https://github.com/hazelcast/hazelcast/commit/5c8bc2c644597f664ee64e3b8b87e1935f2a0439) by me. I was trying to remove `com.hazelcast.instance.BuildInfoProvider.readBuildPropertiesClass(Class<?>, BuildInfo, Overrides)#distribution` and this was leftover from that - but good spot, it's not needed.
> 
> > If you want to reproduce the error, I am sharing a sample POC project.
> > https://github.com/omercelikceng/native-poc-cache
> 
> Thanks for the reproducer, but it seems to work for me locally?
> 
> ```
> [INFO] Successfully built image 'docker.io/library/native-poc-cache:0.0.1-SNAPSHOT'
> [INFO] 
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD SUCCESS
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  04:18 min
> [INFO] Finished at: 2024-08-01T14:03:10+01:00
> [INFO] ------------------------------------------------------------------------
> ```
> 
> Could this be platform-specific?

When you run the code, it throws an error. In other words, it throws an error when the project is starting up.

**Comment by omercelikceng** [comment]
> `artifactId` variable introduced in [5c8bc2c](https://github.com/hazelcast/hazelcast/commit/5c8bc2c644597f664ee64e3b8b87e1935f2a0439) by me. I was trying to remove `com.hazelcast.instance.BuildInfoProvider.readBuildPropertiesClass(Class<?>, BuildInfo, Overrides)#distribution` and this was leftover from that - but good spot, it's not needed.
> 
> > If you want to reproduce the error, I am sharing a sample POC project.
> > https://github.com/omercelikceng/native-poc-cache
> 
> Thanks for the reproducer, but it seems to work for me locally?
> 
> ```
> [INFO] Successfully built image 'docker.io/library/native-poc-cache:0.0.1-SNAPSHOT'
> [INFO] 
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD SUCCESS
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  04:18 min
> [INFO] Finished at: 2024-08-01T14:03:10+01:00
> [INFO] ------------------------------------------------------------------------
> ```
> 
> Could this be platform-specific?

When you run the code, it throws an error. In other words, it throws an error when the project is starting up.

**Comment by JackPGreen** [comment]
> No, it is not platform-specific. When you run the native image, you get an error. In other words, it throws an error when the project is starting up.

Seems to work for me with `java -jar target/native-poc-cache-0.0.1-SNAPSHOT.jar`

**Comment by omercelikceng** [comment]
> > No, it is not platform-specific. When you run the native image, you get an error. In other words, it throws an error when the project is starting up.
> 
> Seems to work for me with `java -jar target/native-poc-cache-0.0.1-SNAPSHOT.jar`

Can you run the project with docker run and check if it works correctly? The jar is not a native image. This is why you're not encountering errors. You wouldn't get any errors if you run the jar version of the project.

**Comment by omercelikceng** [comment]
> > Can you run the project with **docker run** and check if it works correctly?
> 
> Are you able to provide instructions on how to do that, please?

Ok, thank you. 
First, I want you to check if the image we obtained is a native image. Use the `docker images` command for that.
From the list of images that appears, my image should be around **100 MB**. If the image is significantly larger, we need to check a few things.

If it is around 100 MB, follow the next steps.
First, we need to start a Hazelcast instance with the name hz1. Use this command for that.

`docker run --name hz1 -p 5701:5701 --network=cache-network hazelcast/hazelcast:5.5.0-jdk17`

Then, we need to start the project. Use this command for that.

`docker run --network=cache-network docker.io/library/native-poc-cache:0.0.1-SNAPSHOT`

**Review by JackPGreen — APPROVED** [review]

**Comment by omercelikceng** [comment]
@JackPGreen If you are available, could you also take a look at my other pull request? The two pull requests are related. We worked together up to a certain point. I can explain the issue to you more easily. I can immediately explain the issue to you and show the error with my example code.
Thank you very much.
https://github.com/hazelcast/hazelcast/pull/26387

**Pull request closed without merging**